orttraining/orttraining/training_ops/cpu/rnn/lstm_io_utils.cc

```diff
@@ -0,0 +1,358 @@
+  Tensor* CAll = context->Output(3, CAll_shape);                                       // all cell states
+  const int64_t hidden_size_x4 = static_cast<int64_t>(4) * static_cast<int64_t>(hidden_size);
+  const TensorShape iofc_shape{sequence_length, directions, batch_size, hidden_size_x4};  // [seq_length, directions, batch_size, 4 * hidden_size]
+  Tensor* IOFC = context->Output(4, iofc_shape);                                          // iofc gate computations
+
+  AllocatorPtr alloc;
+  ORT_THROW_IF_ERROR(context->GetTempSpaceAllocator(&alloc));
+
+  ORT_ENFORCE(HAll, "All hidden states output is required for LSTMTraining to compute gradients.");
+  all_hidden_states = HAll->MutableDataAsSpan<T>();
+  const size_t final_state_size = static_cast<size_t>(directions) * static_cast<size_t>(batch_size) *
+                                  static_cast<size_t>(hidden_size);
+  final_hidden_state = Ht ? Ht->MutableDataAsSpan<T>()
+                          : rnn::detail::Allocate(alloc, final_state_size,
+                                                  h_final_ptr_, true, static_cast<T>(0));
+  final_cell_state = Ct ? Ct->MutableDataAsSpan<T>()
+                        : rnn::detail::Allocate(alloc, final_state_size,
+                                                c_final_ptr_, true, static_cast<T>(0));
+
+  ORT_ENFORCE(CAll, "All cell states output is required for LSTMTraining to compute gradients.");
+  const size_t all_states_size = static_cast<size_t>(sequence_length) * static_cast<size_t>(directions) *
+                                 static_cast<size_t>(batch_size) * static_cast<size_t>(hidden_size);
+  all_cell_states = CAll ? CAll->MutableDataAsSpan<T>()
+                         : rnn::detail::Allocate(alloc, all_states_size,
+                                                 call_ptr_, true, static_cast<T>(0));
+
+  ORT_ENFORCE(IOFC, "i, o, f, c gate computation output is required for LSTMTraining to compute gradients.");
+  const size_t iofc_size = static_cast<size_t>(sequence_length) * static_cast<size_t>(directions) *
+                           static_cast<size_t>(batch_size) * static_cast<size_t>(hidden_size) * 4U;
+  iofc = IOFC ? IOFC->MutableDataAsSpan<T>()
+              : rnn::detail::Allocate(alloc, iofc_size, iofc_ptr_, true, static_cast<T>(0));
+}
+
+template <typename T>
+LSTMGradInputs<T>::LSTMGradInputs(OpKernelContext* context, const int directions, const int hidden_size) {
+  // Gradient inputs to the gradient kernel
+  const Tensor* X = context->Input<Tensor>(0);           // input sequence [seq_length, batch_size, input_size]
+  const Tensor* W = context->Input<Tensor>(1);           // weights [directions, 4 * hidden_size, input_size]
+  const Tensor* R = context->Input<Tensor>(2);           // recurrence weights [directions, 4 * hidden_size, hidden_size]
+  const Tensor* B = context->Input<Tensor>(3);           // bias [directions, 8 * hidden_size]
```

Comment:
B looks not used by the bw compute.

Reply:
You're right. B and P are not used for backward computation. 

I think it can be removed from the bw op inputs altogether since none of the gradient computation require them.
Thanks.